README
check_mailbox_health Nagios Plugin README
---------------------

This plugin is used to count mails meeting specific criteria

* For instructions on installing this plugin for use with Nagios,
  see below. In addition, generic instructions for the GNU toolchain
  can be found in the INSTALL file.

* For major changes between releases, read the CHANGES file.

* For information on detailed changes that have been made,
  read the Changelog file.

* This plugin is self documenting.  All plugins that comply with
  the basic guidelines for development will provide detailed help when
  invoked with the '-h' or '--help' options.

You can check for the latest plugin at:
  http://labs.consol.de

Send mail to gerhard.lausser@consol.de for assistance.
Please include the OS type/version and the Perl modules' version
that you are using.
Also, run the plugin with the '-vvv' option and provide the resulting 
version information.  Of course, there may be additional diagnostic information
required as well.  Use good judgment.

For patch submissions and bug reports, please send me a mail. You can also find
me at http://www.nagios-portal.de

Requirements
--------------------------------------------------------
You need to install the Perl module Net::IMAP:Simple first.


How to "compile" the check_mailbox_health script.
--------------------------------------------------------

1) Run the configure script to initialize variables and create a Makefile, etc.

	./configure --prefix=BASEDIRECTORY --with-nagios-user=SOMEUSER --with-nagios-group=SOMEGROUP --with-perl=PATH_TO_PERL --with-statefiles-dir=STATE_PATH

   a) Replace BASEDIRECTORY with the path of the directory under which Nagios
      is installed (default is '/usr/local/nagios')
   b) Replace SOMEUSER with the name of a user on your system that will be
      assigned permissions to the installed plugins (default is 'nagios')
   c) Replace SOMEGRP with the name of a group on your system that will be
      assigned permissions to the installed plugins (default is 'nagios')
   d) Replace PATH_TO_PERL with the path where a perl binary can be found.
      Besides the system wide perl you might have installed a private perl
      just for the nagios plugins (default is the perl in your path).
   e) Replace STATE_PATH with the directory where you want the script to
      write state files which transport information from one run to the next.
      (default is /tmp)

   Simply running ./configure will be sufficient to create a check_mailbox_health
   script which you can customize later.
      

2) "Compile" the plugin with the following command:

	make

    This will produce a "check_mailbox_health" script. You will also find
    a "check_mailbox_health.pl" which you better ignore. It is the base for
    the compilation filled with placeholders. These will be replaced during
    the make process.


3) Install the compiled plugin script with the following command:

	make install

   The installation procedure will attempt to place the plugin in a 
   'libexec/' subdirectory in the base directory you specified with
   the --prefix argument to the configure script.


4) Verify that your configuration files for Nagios contains
   the correct paths to the new plugin.


Command line parameters
-----------------------

--hostname=<the mail server>
   This is the hostname or the ip-address.

--port=<the port>
   This is the port where an instance listens. Default: 1433

--mailbox=<mailbox>
   This is an optional mailbox name

--user=<username>
   This is the user which reads the mailbox

--password=<secret>
   This is the user's password.

--mode=<operation mode>
   This parameter tells the plugin what it should check.
   The list of known modes may grow frequently. Please look at 
   http://www.consol.com/opensource/nagios/check-mailbox-health for a list
   of features.

--warning=<warning threshold>
  If the metric is out of this range, the plugin returns a warning.

--critical=<critical threshold>
  If the metric is out of this range, the plugin returns a critical.

   
--------------------------------------
That's it.  If you have any problems or questions, feel free to send mail
to gerhard.lausser@consol.de


ChangeLog
* 2022-08-19 2.0.0.1
  dreckstippfehler
* 2022-08-19 2.0
  support outlook365 with OAuth 2.0 client credentials flow
* 2021-05-04 1.9
  check imap capabilities first, don't try to sort by default
* 2021-01-31 1.8
  reduce debugging output, better handling of utf8 wide characters
* 2017-05-11 1.7.2.1
  update GLPlugin
* 2017-04-20 1.7.2
  filter attachments by file extension (Thanks Philip)
* 2016-11-28 1.7.1
  fix utf-encoded subjects
* 2016-08-10 1.7
  delegate some selects to the imap-server
  add a greenmail simulator for tests
* 2016-08-02 1.6.0.2
  update GLPlugin & bugfix in list-mails
* 2016-06-03 1.6.0.1
  update GLPlugin
* 2016-05-14 1.6
  add selectors "seen" and "is_bigger"
* 2016-02-29 1.5
  update version, changelog, glplugin
* 2016-02-09 1.4
  add glplugin submodule
* 2015-07-13 1.3
  new directory layout, standalone
* 2015-06-02 1.2
  add morph-message
  add date:manip strings for older_than newer_than
* 2015-06-01 1.1
  start with --disable-standalone
* 2015-05-30 1.0
  init


plugins-scripts/Classes/IMAP.pm
package Classes::IMAP;
our @ISA = qw(Classes::Mailserver);

use strict;
use Time::HiRes;
use IO::File;
use File::Copy 'cp';
use Data::Dumper;


sub check_connect_and_version {
  my $self = shift;
  $self->{tic} = time;
  if (! $self->opts->port && ! $self->opts->ssl) {
    $self->override_opt('port', 143);
  } elsif (! $self->opts->port) {
    $self->override_opt('port', 993);
  } else {
    $self->override_opt('hostname', $self->opts->hostname.':'.$self->opts->port);
  }
  my %options = ();
  $options{timeout} = $self->opts->timeout;
  $options{use_ssl} = 1 if $self->opts->ssl;
  $options{debug} = $self->opts->verbose if $self->opts->verbose > 10;
  $self->{session} = Net::IMAP::Simple->new(
      $self->opts->hostname, %options);
  #$self->{session}->starttls() if $self->opts->ssl;
  if (! $self->{session}) {
    $self->add_critical("imap server not available");
  } elsif (! $self->{session}->login(
      $self->opts->username, $self->opts->password)) {
    $self->add_critical("could not open imap3 mailbox (user or password probably wrong)");
  }
  $self->{tac} = time;
  if ($self->{tac} - $self->{tic} >= $self->opts->timeout) {
    $self->add_unknown("timeout");
  }
}

sub check_folder {
  my $self = shift;
  $self->{mails} = [];
  my $msgnums = $self->{session}->select($self->opts->folder);
  $msgnums = 0 if (! defined $msgnums || $msgnums eq "0E0");
  $self->{num_all_mails} = $msgnums;
  $self->debug(sprintf 'there are %d messages in folder %s',
      $msgnums, $self->opts->folder);
  if ($self->{session}->errstr()) {
    $self->debug(sprintf 'error is: %s', $self->{session}->errstr());
    $self->add_unknown($self->{session}->errstr());
    return;
  }
  return;
}

sub read_mails {
  my $self = shift;
}

sub filter_mails {
  my ($self) = @_;
  # --select subject='Anlieferung.+aktuell' --select attachments=text/pdf
  my @filters = ();
  if ($self->opts->select) {
    # Vorselektieren, aber mit Fingerspitzengefuehl, also auf newer/older_than
    # einen Tag zur Sicherheit draufzaehlen.
    # Das beschleunigt die Abfrage ungemein, entlastet das Monitoring
    # und laesst den IMAP-Server die Arbeit tun.
    foreach my $key (keys %{$self->opts->select}) {
      if (lc $key eq "newer_than" || lc $key eq "older_than") {
        my $when = $self->opts->select->{$key};
        if ($when =~ /^\d+$/) {
          if (lc $key eq "newer_than") {
            $when = sprintf "%d minutes ago", $when;
          } else {
            $when = sprintf "in %d minutes", $when;
          }
        }
        my $date = new Date::Manip::Date;
        my $stderrvar;
        my $parseerr;
        *SAVEERR = *STDERR;
        open OUT ,'>',\$stderrvar;
        *STDERR = *OUT;
        eval {
          $parseerr = $date->parse($when);
          my $security = new Date::Manip::Delta;
          if (lc $key eq "newer_than") {
            $security->parse('1 day ago');
          } else {
            $security->parse('in 1 day');
          }
          $date = $date->calc($security);
        };
        if ($@ || $stderrvar || $parseerr) {
          $self->add_unknown(sprintf "invalid date format \"%s\" (%s)",
              $self->opts->select->{$key}, $@ || $stderrvar || $parseerr);
        } else {
          if (lc $key eq "newer_than") {
            push(@filters, sprintf 'SINCE %s', $date->printf("%d-%b-%Y"));
          } else {
            push(@filters, sprintf 'BEFORE %s', $date->printf("%d-%b-%Y"));
          }
        }
        *STDERR = *SAVEERR;
      } elsif (lc $key eq "bigger_than") {
        my $size = $self->opts->select->{$key};
        $self->override_opt('units', 'MB') if ! $self->opts->units();
        if (lc $self->opts->units eq "kb") {
          $size *= 1024;
        } elsif (lc $self->opts->units eq "mb") {
          $size *= (1024 * 1024);
        } elsif (lc $self->opts->units eq "gb") {
          $size *= (1024 * 1024 * 1024);
        }
        push(@filters, sprintf 'LARGER %d', $size * 0.8);
      } elsif (lc $key eq "seen") {
        push(@filters, 'SEEN') if $self->opts->select->{$key} eq "yes";
        push(@filters, 'UNSEEN') if $self->opts->select->{$key} eq "no";
      }
    }
  }
  my @msgids = ();
  $self->getcapabilities();
  if (@filters and $self->can_sort()) {
    @msgids = ($self->{session}->search(join(" ", @filters), "ARRIVAL"));
  } else {
    @msgids = (1..$self->{num_all_mails});
  }
  foreach my $msgnum (@msgids) {
    my $seen = $self->{session}->seen($msgnum);
    my $mail = $self->{session}->get($msgnum);
    next if ! defined $mail; # notbremse, evt. wurde gerade eine mail geloescht
    $self->{session}->unsee($msgnum) if ! $seen;
    my $mailsig = "-new-";
    eval {
      my $mail = Classes::MAIL->new(join("", @{$mail}));
      $mail->{seen} = $seen;
      push(@{$self->{mails}}, $mail) if ! $mail->is_spam;
    };
    if ($@) {
      $self->add_unknown($@);
    }
  }
  $self->SUPER::filter_mails();
}
 
sub getcapabilities {
  my ($self) = @_;
  $self->{capabilities} = [];
  eval {
    my @lines;
    $self->{session}->_process_cmd(
        cmd => [ "CAPABILITY" ],
        final => sub {
            my $cap_line = join( ' ', @lines);
            my @caps = split(/\s+/, $cap_line);
            if ($caps[0] eq '*') { shift @caps; }
            $self->{capabilities} = \@caps;
        },
        process => sub {
            push @lines, @_;
        },
    );
  };
  $self->debug(sprintf "capabilities: %s", join(",", @{$self->{capabilities}}));
}

sub can_sort {
  my ($self) = @_;
  return (grep /^SORT/, @{$self->{capabilities}}) ? 1 : 0;
}


plugins-scripts/Classes/MAIL.pm
package Classes::MAIL;
our @ISA = qw(Monitoring::GLPlugin::TableItem);
use strict;
use List::MoreUtils qw(natatime);
use Date::Manip;
use Encode qw(decode);
our $AUTOLOAD;

sub new {
  my $class = shift;
  my $raw_text = shift;
  my $self = {
    attachments => [],
  };
  bless $self, $class;
  my $stderrvar;
  *SAVEERR = *STDERR;
  open OUT ,'>',\$stderrvar;
  *STDERR = *OUT;
  my $parsed = undef;
  eval {
    $parsed = Email::MIME->new($raw_text);
  };
  if ($@) {
    $self->add_unknown(sprintf "scheise mail %s", $raw_text);
  }
  *STDERR = *SAVEERR;
  if ($stderrvar) {
    if ($stderrvar =~ /Illegal Content-Type parameter/) {
    } elsif ($stderrvar =~ /Use of uninitialized value in lc.*Header.pm/) {
    } else {
      printf STDERR "there is output on stderr, increase verbosity to > 15"
          if $self->opts->verbose < 15;
      if ($self->opts->verbose >= 15) {
        printf STDERR "stderror found\n";
        printf STDERR "stderrvar %s\n", $stderrvar;
        printf STDERR "stderrvar %s\n----------------------------------\n", $raw_text;
      }
    }
    $stderrvar = undef;
  }
  my $header_values = {};
  my @header_names = $parsed->header_pairs;
  my $iter = natatime(2, @header_names);
  while (my @vals = $iter->()) {
    next if $vals[0] eq "Received" && exists $header_values->{received}; # an oberster Stelle ist der letzte Hop
    $header_values->{lc $vals[0]} = $vals[1]; # weil manchmal Message-ID, manchmal Message-Id
  }
  my %save_header_values = %{$header_values};
  *SAVEERR = *STDERR;
  open OUT ,'>',\$stderrvar;
  *STDERR = *OUT;
  eval {
    my $date = new Date::Manip::Date;
    if (! exists $header_values->{received}) {
      # vom lokalen Notes versandt
      my $err = $date->parse($header_values->{date});
      $header_values->{received} = $date->printf("%s");
      #$header_values->{received} = $date->printf("%Y/%m/%d %H:%M:%S");
    } else {
      $header_values->{received} =~ s/.*;//;
      my $err = $date->parse($header_values->{received});
      $header_values->{received} = $date->printf("%s");
      #$header_values->{received} = $date->printf("%Y/%m/%d %H:%M:%S");
    }
  };
  *STDERR = *SAVEERR;
  if ($@ || $stderrvar) {
    if (! $header_values->{received} && exists $header_values->{'x-olkeid'}) {
      $header_values->{subject} = '_check_mailbox_health_SPAM_';
    } elsif ($stderrvar =~ /Illegal Content-Type parameter/) {
      # z.b.
      # Illegal Content-Type parameter ; name=stammtisch.ics at /usr/share/perl5/Email/MIME.pm line 34
      # bei
      # Content-Type: text/calendar;method=PUBLISH;charset=UTF-8;; name=stammtisch.ics
      # Content-Disposition: inline,filename=stammtisch.ics; filename=stammtisch.ics
      # das raeumt dann die content_type-Methode auf
    } elsif ($stderrvar =~ /Use of uninitialized value in lc at .*Header\.pm/) {
    } else {
      printf STDERR "there is output on stderr, increase verbosity to > 15"
          if $self->opts->verbose < 15;
      if ($self->opts->verbose >= 15) {
        printf STDERR "2noreceived %s\n%s\n", $stderrvar, Data::Dumper::Dumper($header_values);
        printf STDERR "3noreceived %s\n%s\n", $stderrvar, $raw_text;
        printf "--------------------------------------------------------------------------------\n\n";
      }
      $self->add_unknown(sprintf "date error in %s from %s",
          $header_values->{'message-id'}, $header_values->{from});
    }
  }
  %{$self->{header_values}} = %{$header_values};
  $self->{body} = $parsed->body;
  #my @ attachments = $parsed->subparts;
  # flatten
  @{$self->{attachments}} = $parsed->subparts;
  foreach (@{$self->{attachments}}) {
    bless $_, "Classes::Attachment";
    $_->parse_attachments();
  }
  return $self;
}

sub is_spam {
  my $self = shift;
  return $self->subject eq "_check_mailbox_health_SPAM_" ? 1 : 0;
}

sub signature {
  my $self = shift;
  return sprintf "%s %s %s\n%s",
      scalar localtime $self->{header_values}->{received},
      $self->{header_values}->{from},
      $self->{header_values}->{subject},
      join("\n", map {
          "  ".$_->content_type()." ".$_->filename();
      } @{$self->{attachments}});
}

sub dump {
  my $self = shift;
  printf "[MESSAGE_%s]\n", $self->message_id();
  printf "Received: %s\n", $self->received();
  printf "From: %s\n", $self->from();
  printf "Subject: %s\n", $self->subject();
#printf "Header %s\n", Data::Dumper::Dumper($self->{header_values});
  my $sisi = $self->size();
  printf "size %.2fKB final\n", $sisi / 1024;
  printf "\n";
}

sub age_minutes {
  my $self = shift;
  return (time - $self->{header_values}->{received}) / 60;
}

sub received {
  my $self = shift;
  return $self->{header_values}->{received};
}

sub message_id {
  my $self = shift;
  return $self->{header_values}->{'message-id'};
}

sub from {
  my $self = shift;
  return $self->{header_values}->{from};
}

sub subject {
  my $self = shift;
  return decode("MIME-Header", $self->{header_values}->{subject} || "");
}

sub body {
  my $self = shift;
  return $self->{body} || "";
}

sub num_attachments {
  my $self = shift;
  return scalar(@{$self->{attachments}});
}

sub attachments {
  my $self = shift;
  return @{$self->{attachments}};
}

sub size {
  my $self = shift;
  my $size = 0;
  $size += length($_->body());
  foreach (@{$self->{attachments}}) {
    my $asize = $_->size();
    $size += $asize;
  }
  return $size;
}

sub raw_size {
  my $self = shift;
  my $size = 0;
  $size += length($_->body_raw());
  foreach (@{$self->{attachments}}) {
    my $asize = $_->size();
    $size += $asize;
  }
  return $size;
}


package Classes::Attachment;
our @ISA = qw(Email::MIME Classes::MAIL Monitoring::GLPlugin::TableItem);
use strict;

sub new {
  my $class = shift;
  my $raw_text = shift;
  my $self = {
    attachments => [],
  };
  bless $self, $class;
  my $parsed = undef;
  eval {
    $parsed = Email::MIME->new($raw_text);
    
    @{$self->{attachments}} = $parsed->subparts;
    $self->{body} = $parsed->body;
  };
  if ($self->content_type !~ /^text/) {
    $self->body_set("_binaerer_schrott_");
  }
  return $self;
}

sub parse_attachments {
  my $self = shift;
  $self->{attachments} = [];
  my @attachments = $self->subparts();
  foreach (@attachments) {
    bless $_, "Classes::Attachment";
    $_->parse_attachments();
    push(@{$self->{attachments}}, $_);
  }
}

sub filename {
  my $self = shift;
  my $filename = "--noname--";
  if (defined $self->SUPER::filename) {
    $filename = $self->SUPER::filename;
    $filename =~ s/;.*//g;
    # irgend so utfdecodierte filenamen ?utf?schlotznrotzn= kreuzen mit
    # anfuehrungszeichen auf.
    $filename =~ s/"$//g;
    $filename =~ s/^"//g;
  }
  return $filename
}

sub content_type {
  my $self = shift;
  my $content_type = "text/plain";
  if (! defined $self->SUPER::content_type) {
    if (exists $self->{ct} && exists $self->{ct}->{type} && exists $self->{ct}->{subtype}) {
      $content_type = $self->{ct}->{type}."/".$self->{ct}->{subtype};
    } else {
      $content_type = "text/plain";
    }
  } else {
    $content_type = $self->SUPER::content_type;
    $content_type =~ s/;.*//g;
  }
  return $content_type
}



plugins-scripts/Classes/O365.pm
package Classes::O365;
our @ISA = qw(Classes::IMAP);

use strict;
use Time::HiRes;
use IO::File;
use File::Copy 'cp';
use MIME::Base64;
use JSON;
use Data::Dumper;


sub aquire_auth_token {
  # https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow
  # No, i will not help you setting up the azure environment.
  # You will have to bother your own admins. You can also ask Microsoft.
  my $self = shift;
  my $client_secret = $self->opts->password;
  my ($tenant_id, $client_id) = split("/", $self->opts->username);
  my $ua = LWP::UserAgent->new();
  $ua->agent(qq{Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0});
  my $scope = "https://outlook.office365.com/.default";
  my $uri = "https://login.microsoftonline.com/".$tenant_id."/oauth2/v2.0/token";
  my $header = ['Content-Type' => 'application/x-www-form-urlencoded'];
  my $data =
      "client_id=$client_id".
      "&client_secret=$client_secret".
      "&scope=$scope".
      "&grant_type=client_credentials";
  my $r = HTTP::Request->new('POST', $uri, $header, $data);
  my $res = $ua->request($r);
  $self->debug(sprintf "token request is %s", Data::Dumper::Dumper($r));
  $self->debug(sprintf "token request uri is %s", $uri);
  my @token = ();
  eval {
    my $message = $res->decoded_content;
    my $fromjson = from_json($message);
    $self->debug(sprintf "token response is %s", Data::Dumper::Dumper($fromjson));
    @token = ($fromjson->{access_token}, time + $fromjson->{expires_in});
  };
  if ($@) {
    @token = (undef, undef);
  }
  return @token;
}

sub check_connect_and_version {
  my $self = shift;
  $self->{tic} = time;
  if (! $self->opts->port && ! $self->opts->ssl) {
    $self->override_opt('port', 143);
  } elsif (! $self->opts->port) {
    $self->override_opt('port', 993);
  } else {
    $self->override_opt('hostname', $self->opts->hostname.':'.$self->opts->port);
  }
  my %options = ();
  $options{timeout} = $self->opts->timeout;
  $options{use_ssl} = 1 if $self->opts->ssl;

  {
    #
    # username = tenant_id/client_id/user@domain
    # password = client_secret
    # tokens aus dem cache holen (refresh und access)
    # pruefen, ob man sich mit dem _token einen xoauth2-token holen kann
    # falls nicht, den ganzen api-schlonz ausfuehren
    no warnings 'redefine';
    no warnings 'once';
    *Net::IMAP::Simple::login = sub {
      my ($self, $o365, $username, $password) = @_;
      my $client_secret = $password;
      my ($tenant_id, $client_id, $mailbox_user) = split("/", $username);
      # weil wir hier in classify() sind, ein fehlendes statefilesdir aber
      # erst im drauffolgenden validate_args() auf einen default gesetzt wird,
      # muss der code teilweise hierher vorgezogen werden.
      if ($o365->opts->can("statefilesdir") && ! $o365->opts->statefilesdir) {
        if ($^O =~ /MSWin/) {
          if (defined $ENV{TEMP}) {
            $o365->override_opt('statefilesdir',
                $ENV{TEMP}."/".$Monitoring::GLPlugin::plugin->{name});
          } elsif (defined $ENV{TMP}) {
            $o365->override_opt('statefilesdir',
                $ENV{TMP}."/".$Monitoring::GLPlugin::plugin->{name});
          } elsif (defined $ENV{windir}) {
            $o365->override_opt('statefilesdir',
                File::Spec->catfile($ENV{windir}, 'Temp')."/".$Monitoring::GLPlugin::plugin->{name});
          } else {
            $o365->override_opt('statefilesdir',
                "C:/".$Monitoring::GLPlugin::plugin->{name});
          }
        } elsif (exists $ENV{OMD_ROOT}) {
          $o365->override_opt('statefilesdir',
              $ENV{OMD_ROOT}."/var/tmp/".$Monitoring::GLPlugin::plugin->{name});
        } else {
          $o365->override_opt('statefilesdir',
              "/var/tmp/".$Monitoring::GLPlugin::plugin->{name});
        }
      }
      $Monitoring::GLPlugin::plugin->{statefilesdir} =
          $o365->opts->statefilesdir if $o365->opts->can("statefilesdir");
      $Monitoring::GLPlugin::mode = (
          map { $_->{internal} }
          grep {
             ($o365->opts->mode eq $_->{spec}) ||
             ( defined $_->{alias} && grep { $o365->opts->mode eq $_ } @{$_->{alias}})
          } @{$Monitoring::GLPlugin::plugin->{modes}}
      )[0];

      my $cache = $o365->load_state(name => "o365_tokens_".$client_id) || {
          auth_token => undef,
      };
      # eventuell validitaet des oauth_token/auth_token ermitteln
      if (! $cache->{auth_token} or $cache->{valid_until} < time - 600) {
        ($cache->{auth_token}, $cache->{valid_until}) = $o365->aquire_auth_token();
        $o365->save_state(name => "o365_tokens_".$client_id, save => {
            client_id => $client_id,
            tenant_id => $tenant_id,
            auth_token => $cache->{auth_token},
            valid_until => $cache->{valid_until},
        }) if $cache->{auth_token};
      }
      if (! $cache->{auth_token}) {
        #$o365->add_critical("unable to aquire a xoauth2 token");
        $self->{_errstr} = "unable to aquire a xoauth2 token";
        return undef;
      }
      my $auth_token = $cache->{auth_token};
      my $oauth_sign = sprintf "user=%s\x01auth=Bearer %s\x01\x01",
          $mailbox_user, $auth_token;
      $o365->debug(sprintf "SIGNING IN (%s)\n", $oauth_sign);
      $oauth_sign = encode_base64($oauth_sign, '');
      return $self->_process_cmd(
          cmd     => [ "AUTHENTICATE XOAUTH2" => qq[$oauth_sign] ],
          final   => sub { 1 },
          process => sub { },
      );
    };
  }

  $self->{session} = Net::IMAP::Simple->new(
      $self->opts->hostname, %options);
  #$self->{session}->starttls() if $self->opts->ssl;
  if (! $self->{session}) {
    $self->add_critical("imap server not available");
  } else {
    my $login = $self->{session}->login(
        $self, $self->opts->username, $self->opts->password);
    if (! $login) {
      $self->add_critical($self->{session}->errstr);
    }
  }
  $self->{tac} = time;
  if ($self->{tac} - $self->{tic} >= $self->opts->timeout) {
    $self->add_unknown("timeout");
  }
}



plugins-scripts/Classes/Mailserver.pm
package Classes::Mailserver;
our @ISA = qw(Classes::Device);

use Encode qw(encode_utf8 is_utf8 decode_utf8);
use strict;

sub init {
  my $self = shift;
  if ($self->mode =~ /^server::connectiontime/) {
    $self->{connection_time} = $self->{tac} - $self->{tic};
  } elsif ($self->mode =~ /^mails::/) {
    $self->check_folder();
    return if $self->check_messages();
    $self->{num_all_mails} = scalar(@{$self->{mails}}) if ! exists $self->{num_all_mails};
    $self->filter_mails();
    $self->{num_mails} = scalar(@{$self->{mails}});
    if ($self->mode =~ /^mails::age/) {
      $self->{max_mail_age} = 0;
      $self->{num_old_mails} = 0;
      map {
        $self->{max_mail_age} = $_->age_minutes if $_->age_minutes > $self->{max_mail_age};
      } @{$self->{mails}};
    }
  } else {
    $self->no_such_mode();
  }
  $self->check();
  $self->dump()
      if $self->opts->verbose >= 2;
}

sub dump {
  my $self = shift;
  foreach (@{$self->{mails}}) {
    $_->dump();
  }
}

sub check {
  my $self = shift;
  if ($self->mode =~ /^server::connectiontime/) {
    $self->set_thresholds(warning => 1, critical => 5);
    $self->add_message($self->check_thresholds($self->{connection_time}),
         sprintf "%.2f seconds to connect as %s",
              $self->{connection_time}, $self->opts->username,);
    $self->add_perfdata(
        label => 'connection_time',
        value => $self->{connection_time},
    );
  } elsif ($self->mode =~ /^mails::list/) {
    foreach (@{$self->{mails}}) {
      printf "%s\n", $_->signature;
    }
    $self->add_ok("have fun");
  } elsif ($self->mode =~ /^mails::/) {
    if ($self->opts->select) {
      if (grep /attachments/, keys %{$self->opts->select}) {
        $self->add_info(sprintf "%d mails with attachments selected (%d total in mailbox)", $self->{num_mails}, $self->{num_all_mails});
      } else {
        $self->add_info(sprintf "%d mails selected (%d total in mailbox)", $self->{num_mails}, $self->{num_all_mails});
      }
    } else {
      $self->add_info(sprintf "%d mails in mailbox", $self->{num_mails});
    }
    if ($self->mode =~ /^mails::age/) {
      $self->add_ok(); # statusmeldung von da oben
      $self->set_thresholds(metric => 'max_age', warning => 30, critical => 60);
      if ($self->{num_mails}) {
        # alle selektierten mails, unter diesen wurde max_mail_age ermittelt
        if ($self->check_thresholds(metric => 'max_age', value => $self->{max_mail_age})) {
          # wenn die aelteste mail zu alt ist
          @{$self->{mails}} = sort {
            $a->age_minutes <=> $b->age_minutes;
          } @{$self->{mails}};
          @{$self->{mails}} = grep {
            # hole alle mails, die aelter als die thresholds sind
            $self->check_thresholds(metric => 'max_age', value => $_->age_minutes);
          } @{$self->{mails}};
          $self->{num_mails} = scalar(@{$self->{mails}});
          if (scalar(@{$self->{mails}})) {
            $self->add_message($self->check_thresholds(metric => 'max_age', value => $self->{max_mail_age}),
                sprintf "%d mails too old", $self->{num_mails});
            $self->add_info(sprintf "age of the oldest mail is %d minutes", $self->{max_mail_age});
          } else {
            $self->add_critical("jetz klauns da de mails scho mittn ausm plugin aussa");
          }
          $self->add_message($self->check_thresholds($self->{max_mail_age}));
        } else {
          $self->add_ok(sprintf "age of the oldest mail is %d minutes", $self->{max_mail_age});
          $self->add_ok("no outdated mails");
        }
      } else {
        $self->add_ok("no mails");
      }
      $self->add_perfdata(
          label => 'mails',
          value => $self->{num_mails},
      );
      $self->add_perfdata(
          label => 'max_age',
          value => $self->{max_mail_age},
      );
    } elsif ($self->mode =~ /^mails::count/) {
      $self->set_thresholds(metric => 'mails', warning => 30, critical => 60);
      $self->add_message($self->check_thresholds(metric => 'mails', value => $self->{num_mails}));
      $self->add_perfdata(
          label => 'mails',
          value => $self->{num_mails},
      );
    }
  }
}

sub num_mails {
  my $self = shift;
  return scalar(@{$self->{mails}});
}

sub filter_mails {
  my $self = shift;
  # --select subject='Anlieferung.+aktuell' --select attachments=text/pdf
  return if ! $self->opts->select;
  my @filters = ();
  foreach my $key (keys %{$self->opts->select}) {
    if (lc $key eq "subject") {
      push(@filters, sub {
        my $self = shift;
        my $mail = shift;
        return $self->filter_namex($self->opts->select->{$key}, $mail->subject);
      });
    } elsif (lc $key eq "content") {
      push(@filters, sub {
        my $self = shift;
        my $mail = shift;
        my $wanted_content = $self->opts->select->{$key};
        if (! $mail->num_attachments) {
          if (is_utf8($mail->body)) {
            binmode STDOUT, ':utf8';
            $wanted_content = decode_utf8($wanted_content);
          }
          return $self->filter_namex($wanted_content, $mail->body);
        } else {
          grep {
            $_->content_type =~ /^text/;
          } grep {
            my $wanted_content = $self->opts->select->{$key};
            if (is_utf8($_->body)) {
              binmode STDOUT, ':utf8';
              $wanted_content = decode_utf8($wanted_content);
            }
            $self->filter_namex($wanted_content, $_->body);
          } @{$mail->{attachments}};
        }
      });
    } elsif (lc $key eq "newer_than") {
      push(@filters, sub {
        my $self = shift;
        my $mail = shift;
        if ($self->opts->select->{$key} =~ /^\d+$/) {
          $mail->age_minutes <= $self->opts->select->{$key};
        } else {
          my $date = new Date::Manip::Date;
          my $stderrvar;
          my $parseerr;
          *SAVEERR = *STDERR;
          open OUT ,'>',\$stderrvar;
          *STDERR = *OUT;
          eval {
            $parseerr = $date->parse($self->opts->select->{$key});
          };
          if ($@ || $stderrvar || $parseerr) {
            $self->add_unknown(sprintf "invalid date format \"%s\" (%s)",
                $self->opts->select->{$key}, $@ || $stderrvar || $parseerr);
          }
          *STDERR = *SAVEERR;
          $mail->received >= $date->printf("%s");
        }
      });
    } elsif (lc $key eq "older_than") {
      push(@filters, sub {
        my $self = shift;
        my $mail = shift;
        if ($self->opts->select->{$key} =~ /^\d+$/) {
          $mail->age_minutes < $self->opts->select->{$key};
        } else {
          my $date = new Date::Manip::Date;
          my $stderrvar;
          my $parseerr;
          *SAVEERR = *STDERR;
          open OUT ,'>',\$stderrvar;
          *STDERR = *OUT;
          eval {
            $parseerr = $date->parse($self->opts->select->{$key});
          };
          if ($@ || $stderrvar || $parseerr) {
            $self->add_unknown(sprintf "invalid date format \"%s\" (%s)", 
                $self->opts->select->{$key}, $@ || $stderrvar || $parseerr);
          }
          *STDERR = *SAVEERR;
          $mail->received < $date->printf("%s");
        }
      });
    } elsif (lc $key eq "has_attachments") {
      push(@filters, sub {
        my $self = shift;
        my $mail = shift;
        return $mail->num_attachments;
      });
    } elsif (lc $key eq "bigger_than") {
      push(@filters, sub {
        my $self = shift;
        my $mail = shift;
        my $size = $mail->size;
        $self->override_opt('units', 'MB') if ! $self->opts->units();
        if (lc $self->opts->units eq "kb") {
          $size /= 1024;
        } elsif (lc $self->opts->units eq "mb") {
          $size /= (1024 * 1024);
        } elsif (lc $self->opts->units eq "gb") {
          $size /= (1024 * 1024 * 1024);
        }
        return $size >= $self->opts->select->{$key};
      });
    } elsif (lc $key eq "attachments") {
      # --select attachments='application/pdf,*.xls,*.xlsx'
      push(@filters, sub {
        my $self = shift;
        my $mail = shift;
        my @types = map {
            /^\s*(.*?)\s*$/; $1;
        } split(/,/, $self->opts->select->{$key});
        grep {
          my $attachment = $_;
          grep {
            my $type = $_;
            if (index($type, "/") != -1) {
              # application/pdf
              if ($self->opts->regexp) {
                # image/.* -> image\/.*
                # image\/.* -> image\/.*
                $type =~ s/(?<!\\)\//\\\//g;
              }
              $self->filter_namex($type, $attachment->content_type) ;
            } else {
              if (is_utf8($attachment->filename)) {
                # das wird richtig spassig. keine ahnung, wie das richtig
                # geht, aber hier ist das so hingedengelt, dass es im test
                # funktioniert.
                my $dollar = 0;
                if ($type =~ /^(.*)\$$/) {
                  # erst den dollar wegnehmen und dann nach den konvertieren
                  # wieder hinten dranpappen. wenn der dollar mitgeutftet
                  # wird, dann klappts mit dem anchor nicht. (seltsamerweise
                  # matcht dann 'jpg.*$', das ist alles so ein dreck.
                  $type = $1;
                }
                $type = decode_utf8($type);
                $type .= '$' if $dollar;
                # noch was: "Wide character in printf" auf stderr und so zeug.
                # laesst sch hoffentlich vermeiden mit
                binmode STDOUT, ':utf8';
              }
              if ($self->opts->regexp) {
                # \.xlsx
                # match on the whole filename
                $self->filter_namex($type, $attachment->filename) ;
              } else {
                # xlsx
                # match exactly on the extension
                if ($attachment->filename =~ /(.*)\.([^\.]+)$/) {
                    $type eq $2;
                } else {
                    0;
                }
              }
            }
          } @types;
        } @{$mail->{attachments}};
      });
    } elsif (lc $key eq "seen") {
      push(@filters, sub {
        my $self = shift;
        my $mail = shift;
        return 1 if $self->opts->select->{$key} eq "no" && ! $mail->{seen};
        return 1 if $self->opts->select->{$key} eq "yes" &&  $mail->{seen};
        return 0;
      });
    }
  }
  my $filters = scalar(@filters);
  my $ok_filters = 0;
  @{$self->{mails}} = grep {
      $ok_filters = 0;
      foreach my $filter (@filters) {
        $ok_filters++ if
            $filter->($self, $_);
      }
      $ok_filters == $filters;
  } @{$self->{mails}};
}

